External image capture element

miguelaenlle opened the pull request:
Related to #11573

miguelaenlle commented on apps/prairielearn/elements/pl-external-image-capture/pl-external-image-capture.py:
```diff
@@ -0,0 +1,78 @@
+
+    submitted_files = data["submitted_answers"].get("_files", [])
+
+    submitted_file_names = list({x.get("name") for x in submitted_files})
```
Just get the first one and extract its name if the submitted files have length > 0

miguelaenlle commented on apps/prairielearn/elements/pl-external-image-capture/pl-external-image-capture.py:
```diff
@@ -0,0 +1,78 @@
+def parse(element_html: str, data: pl.QuestionData) -> None:
+    element = lxml.html.fragment_fromstring(element_html)
+    answer_name = pl.get_string_attrib(element, "answer-name", "1")
+    file_content = data["submitted_answers"].get(answer_name, None)
```
I don't think this error handling works as intended: double check it

Also, is there a better way to indicate to the student that they must confirm their capture before submitting?

Maybe use the last capture as the submission?

miguelaenlle commented on apps/prairielearn/src/lib/db-types.ts:
```diff
@@ -483,6 +483,18 @@ export const AiQuestionGenerationPromptSchema = z.object({
 
 export type AiQuestionGenerationPrompt = z.infer<typeof AiQuestionGenerationPromptSchema>;
 
+export const ExternalImageCaptureSchema = z.object({
```
Instead of external image capture, what about external submission? Since it's an external submission on tablet/mobile

miguelaenlle commented on apps/prairielearn/src/lib/externalImageCapture.ts:
```diff
@@ -0,0 +1,48 @@
+    resLocals: Record<string, any>;
+}) => {
+    const file_id = await uploadFile({
+        display_filename: 'external-image-capture.png',
```
Should we use different file names for each submission? Or let the user specify this?

miguelaenlle commented on apps/prairielearn/src/lib/externalImageCapture.ts:
```diff
@@ -0,0 +1,48 @@
+        }
+    );
+
+    // Emit a socket event to notify the client that the image has been capture
```
Comment typo

miguelaenlle commented on apps/prairielearn/src/lib/externalImageCaptureSocket.ts:
```diff
@@ -0,0 +1,93 @@
+
+export function connection(socket: Socket) {
+    socket.on('joinExternalImageCapture', async (msg, callback) => {
+        console.log('Msg', msg);
```
Remove conosle logs

miguelaenlle commented on apps/prairielearn/src/lib/externalImageCaptureSocket.ts:
```diff
@@ -0,0 +1,93 @@
+        console.log('Msg', msg);
+        console.log('Callback', callback);
+
+        // TODO: Implement token authentication
```
Remove tODO

miguelaenlle commented on apps/prairielearn/src/lib/externalImageCaptureSocket.ts:
```diff
@@ -0,0 +1,93 @@
+
+        socket.join(`variant-${msg.variant_id}-answer-${msg.answer_name}`);
+
+        const existsImageCapture = await sqldb.queryOptionalRow(
```
Remove the callback

miguelaenlle commented on apps/prairielearn/src/lib/externalImageCaptureSocket.ts:
```diff
@@ -0,0 +1,93 @@
+        namespace.to(`variant-${variant_id}-answer-${answer_name}`).emit('imageUploaded', {
+            variant_id,
+            answer_name,
+            image_uploaded: true,
```
Image_uploaded is always `true` when you remove the initial callback, so just remove it.

miguelaenlle commented on apps/prairielearn/src/lib/externalImageCaptureSocket.ts:
```diff
@@ -0,0 +1,93 @@
+};
+
+// TODO: Create a separate function for this -- this is shared
+function checkToken(token: string, variantId: string): boolean {
```
Complete the TODO

miguelaenlle commented on apps/prairielearn/src/lib/externalImageCaptureSocket.types.ts:
```diff
@@ -0,0 +1,4 @@
+export interface StatusMessage {
```
Never used anywhere: should probably use this for type safety in the serverside sockets

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.html.ts:
```diff
@@ -0,0 +1,88 @@
+export function ExternalImageCapture({ resLocals }: { resLocals: Record<string, any> }) {
+    return PageLayout({
+        resLocals,
+        pageTitle: 'External image capture',
```
Change this name.

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.html.ts:
```diff
@@ -0,0 +1,88 @@
+                    for="camera-input"
+                    class="d-flex flex-column align-items-center justify-content-center
+                        border border-primary rounded p-4 text-center"
+                    style="cursor: pointer; border-style: dashed !important;"
```
Is the !important necessary?

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.html.ts:
```diff
@@ -0,0 +1,88 @@
+                <i class="bi bi-check-circle-fill text-success me-2"></i>
+                Scan successful
+            </h1>
+            <p>Click "Refresh" under the submitted file list in the assessment to update your submission.</p>
```
Update this comment - we have websockets now

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.html.ts:
```diff
@@ -0,0 +1,88 @@
+                        />
+                    </div>
+
+                    <i class="bi bi-camera text-primary" style="font-size: 2rem;"></i>
```
Can we use the built-in fontsize classes for Bootstrap instead?

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.sql:
```diff
@@ -0,0 +1,40 @@
+  $answer_name,
+  $file_id
+) 
+ON CONFLICT ON CONSTRAINT external_image_capture_variant_element_uuid_unique
```
The constraint name is no longer correct -- should be `variant_id_answer_name_unique`

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.sql:
```diff
@@ -0,0 +1,40 @@
+ON CONFLICT ON CONSTRAINT external_image_capture_variant_element_uuid_unique
+  DO UPDATE SET file_id = EXCLUDED.file_id;
+
+-- BLOCK select_external_image_capture_by_variant_and_element
```
`_and_answer_name`

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,191 @@
+router.post(
+    '/answer_name/:answer_name',
+    asyncHandler(async (req, res) => {
+        // Validate that the user has access to the variant
```
Inaccurate comment

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,191 @@
+        // Validate that the user has access to the variant
+        const variantId = req.params.variant_id;
+        const answer_name = req.params.answer_name;
+        const userId = res.locals.authn_user.user_id;
```
Move this down

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,191 @@
+        const answer_name = req.params.answer_name;
+        const userId = res.locals.authn_user.user_id;
+
+        if (!variantId || !answer_name || !userId) {
```
These will never be missing

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,191 @@
+        });
+
+        if (!variant) {
+            res.status(404).send('Variant not found');
```
Use the real status pages?

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,191 @@
+    })
+)
+
+// router.post(
```
Remove this commented out code

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.ts:
```diff
@@ -202,6 +203,12 @@ const PostBodySchema = z.union([
     description: z.string(),
   }),
+  z.object({
+    __action: z.literal('upload_external_image_capture'),
```
Update this name

miguelaenlle commented on apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.ts:
```diff
@@ -244,6 +244,10 @@ router.post(
         `${res.locals.urlPrefix}/instance_question/${res.locals.instance_question.id}/?variant_id=${variant_id}`,
       );
+    } else if (req.body.__action === 'upload_external_image_capture') {
+      // This action is handled in the externalImageCapture middleware, so it
```
Remove this

miguelaenlle commented on apps/prairielearn/src/question-servers/freeform.ts:
```diff
@@ -1304,6 +1304,8 @@ async function renderPanel(
     ? locals.questionUrl + `submission/${submission?.id}/file`
     : null;
 
+  console.log('instance question', locals.instance_question);
```
Remove this

miguelaenlle commented on apps/prairielearn/src/question-servers/freeform.ts:
```diff
@@ -1312,12 +1314,23 @@ async function renderPanel(
     submission_files_url: submission ? submissionFilesUrl : null,
+    submission_date: submission?.date ?? null,
+
+    csrf_token: locals.__csrf_token,
```
I don't think we need to expose the CSRF token anymore

jonatanschroeder commented:
> * Change the name from `pl-external-image-capture` to `pl-handwriting-input`, since submissions aren't necessarily coming from an external device now that we have webcam input support.

What about `pl-image-capture`? It's also not necessarily handwriting.

miguelaenlle commented on testCourse/questions/__drafts__/draft_173/info.json:
```diff
@@ -0,0 +1,6 @@
+{
```
Remove this draft...

eliotwrobson commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.py:
```diff
@@ -0,0 +1,78 @@
+    instance_question_id = data["options"].get("instance_question_id", None)
+
+    if course_instance_id is not None and instance_question_id is not None:
+        external_image_capture_url = f"{data['options'].get('serverCanonicalHost')}/pl/course_instance/{course_instance_id}/instance_question/{instance_question_id}/variants/{html_params['variant_id']}/external_image_capture/answer_name/{html_params['answer_name']}"
```
I think you should use the normal parenthesis lookup for `'serverCanonicalHost'`, since if this not present, the lookup will return 'None' and produce an invalid URL, so it would be better to just throw an exception rather than produce a strange URL with errors.

miguelaenlle replied:
Got it; thank you for the heads up on this, @eliotwrobson . I'll make that change 👍

miguelaenlle replied:
I made this change:

https://github.com/PrairieLearn/PrairieLearn/blob/d0044d9611d68497d3587fdf6ac0249d36e2fd17/apps/prairielearn/elements/pl-image-capture/pl-image-capture.py#L57-L73

reteps commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.py:
```diff
@@ -0,0 +1,100 @@
+    if len(submitted_files) > 0:
+        for file in submitted_files:
+            if file["name"] == f"{answer_name}.png":
+                submitted_file_name = file["name"]
```
Agreed -- I think this whole logic can be simplified to 

```
submitted_file_name = next(file["name"] == f"{answer_name}.png" for file in submitted_files, None)
```

miguelaenlle replied:
Thank you for the suggestions, @eliotwrobson @reteps . I implemented this suggestion with a boolean flag:

https://github.com/PrairieLearn/PrairieLearn/blob/e508aba0ae042884ca1c83c37e2580b05a3e3cc6/apps/prairielearn/elements/pl-image-capture/pl-image-capture.py#L34-L39

I didn't use `next(file["name"] == f"{answer_name}.png" for file in submitted_files, None)` since this resulted in `submitted_file_name` evaluating to `True` or `False` without directly extracting the file name. While this could be modified in a way similar to this to extract the filename:

`next(file["name"] if file["name"] == f"{answer_name}.png" else None for file in submitted_files, None)` 

I opted for the boolean check approach to simplify the implementation.

miguelaenlle commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.mustache:
```diff
@@ -0,0 +1,128 @@
+          class="capture-with-webcam-button btn btn-info"
+        >
+          <i class="bi bi-camera-fill me-1"></i>
+          Capture with webcam
```
I agree with your feedback, @jonatanschroeder . Any thoughts on, "Capture with local camera"? 

I was considering "Capture with connected camera", though that implies that the camera isn't necessarily built-in.

jonatanschroeder replied:
I like "local camera".

nwalters512 commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.mustache:
```diff
@@ -0,0 +1,128 @@
+<script>
+$(function() {
+    new window.PLImageCapture(
+      "{{uuid}}",
+      "{{answer_name}}",
+      "{{external_image_capture_url}}",
+      "{{variant_id}}",
+      "{{submitted_file_name}}",
+      "{{submission_date}}",
+      "{{editable}}",
+      "{{mobile_capture_enabled}}",
+    );
```
Nit, some other elements construct a full JSON object in Python:

https://github.com/PrairieLearn/PrairieLearn/blob/09b9386f2e32621310fb2f730435d555dec39870/apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.py#L161

And then interpolate it into the resulting HTML:

https://github.com/PrairieLearn/PrairieLearn/blob/09b9386f2e32621310fb2f730435d555dec39870/apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.mustache#L22

IMO this is a little cleaner.

jonatanschroeder replied:
To add to this: one advantage of using JSON as a data attribute is that you can more easily handle non-string options, as well as makes it easier to handle options that may have `<>&'"` as part of their content.

nwalters512 commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.mustache:
```diff
@@ -0,0 +1,128 @@
+      </button>
+    </div>
+  </div>
+</div>
```
Nit, missing trailing newline.

nwalters512 commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.mustache:
```diff
@@ -0,0 +1,128 @@
+  <div class="card-header">
+    Image capture
+  </div>
+  <div class="capture-preview-container p-3" {{#submission_files_url}}data-submission-files-url="{{submission_files_url}}"{{/submission_files_url}}>
```
Nit, we've been trying to use class names like `js-*` to indicate when a class is used as a selector in JS vs. for CSS styling. Consider doing the same here, though this is not a blocker.

miguelaenlle replied:
Thank you for the suggestion, @nwalters512. I modified these classes used as selectors for JS to use class names like `js-*`

nwalters512 commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.mustache:
```diff
@@ -0,0 +1,128 @@
+            data-bs-placement="auto"
+            data-bs-title="Capture with mobile device"
+            data-bs-content='
+            <div class="w-100 d-flex flex-column align-items-center">
+              <div id="qr-code-{{uuid}}" class="mb-3 bg-body-secondary d-flex justify-content-center align-items-center rounded border" style="width:200px;height:200px">
+                <div class="spinning-wheel spinner-border">
+                  <span class="visually-hidden">Loading...</span>
+                </div>
+              </div>
+              <p class="text-muted mb-0">
+                Scan the QR code with your mobile device to capture an image of your solution. 
+              </p>
+            </div>'
```
Nit (you might consider indenting all the HTML as well):

```suggestion
            <div class="w-100 d-flex flex-column align-items-center">
              <div id="qr-code-{{uuid}}" class="mb-3 bg-body-secondary d-flex justify-content-center align-items-center rounded border" style="width:200px;height:200px">
                <div class="spinning-wheel spinner-border">
                  <span class="visually-hidden">Loading...</span>
                </div>
              </div>
              <p class="text-muted mb-0">
                Scan the QR code with your mobile device to capture an image of your solution. 
              </p>
            </div>
            '
```

nwalters512 commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.py:
```diff
@@ -0,0 +1,106 @@
+
+    pl.check_attribs(
+        element,
+        required_attribs=["answer-name"],
```
We might want to use `file-name` or something instead (check other elements that deal with files to see what might be a good pattern to follow). `answers-name` (note the `s`, so we've already deviated from that a bit) generally indicates a property in `data["submitted_answers"]`, but from what I can see, it's being used to construct a filename with a `.png` suffix.

miguelaenlle replied:
Thank you for pointing this out, @nwalters512 . The `answer-name` is only used to construct a filename with a `.png` suffix. 

Looking at other elements that work with files, it looks like `file-name` is the convention. The only thing is that all elements with a `file-name` attribute also allow users to specify any filename suffix: e.g. `file-name="image.jpg"`. 

Would you have any recommendations on how to address this? There are a few options I see:
- Perform validation on `file-name` to enforce it has a `.png` suffix
- Let users specify `file-name` without the suffix
- Use a different name than `file-name` such as `answers-name`.

jonatanschroeder replied:
Another option is to have a file-name attribute, but allow it to have a default file name, similar to what pl-rich-text-editor does (see #11950).

miguelaenlle replied:
> My recommendation would be to use `file-name` and require users to specify one with a `.png` suffix.

Sounds good; I'll use `file-name` and require users to specify a `png` suffix. I'll implement this functionality and/or any adjustments from further feedback after @eliotwrobson responds.

> Default file name

Thank you for bringing this up, @jonatanschroeder . I think it's a good idea for `pl-image-capture` to have a default filename, particularly when it's used in a question with just one `pl-image-capture` element.

miguelaenlle replied:
I changed `answer-name` to `file-name`, and added suffix validation:

https://github.com/PrairieLearn/PrairieLearn/blob/bc869db050469f0e9f1bcd5baa0a2046c099f2ae/apps/prairielearn/elements/pl-image-capture/pl-image-capture.py#L10-L25

nwalters512 commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.py:
```diff
@@ -0,0 +1,106 @@
+        pl.add_files_format_error(data, f"No image was submitted for {answer_name}.")
+        return
+
+    if not submitted_file_content.startswith("data:"):
+        pl.add_files_format_error(
+            data, f"Image submission for {answer_name} is not a data URI."
+        )
+        return
```
Do we want to validate that the media type is `image/png`? It's not clear to me if that'll be present or not.

Do we want to validate that the file decodes as a `.png` file? We might not care about this.

miguelaenlle replied:
I added validation for the media type: 

https://github.com/PrairieLearn/PrairieLearn/blob/9d52f30e4c3e00ca8f04b56eef17a8537432ecca/apps/prairielearn/elements/pl-image-capture/pl-image-capture.py#L108-L113

To validate that the file decodes as a `.png` file, a library like Python Pillow could be used to load the image and validate its data. I assume this would increase the security of the `pl-image-capture` element and prevent file format-related issues downstream, though it may not be needed. 

What are your thoughts on implementing functionality to ensure that the file decodes as a `png`, @nwalters512 ?

nwalters512 commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.py:
```diff
@@ -0,0 +1,106 @@
+        )
+
+    if mobile_capture_enabled:
+        if course_instance_id is not None and instance_question_id is not None:
+            external_image_capture_url = f"{server_canonical_host}/pl/course_instance/{course_instance_id}/instance_question/{instance_question_id}/variants/{html_params['variant_id']}/external_image_capture/answer/{answer_name}"
+        elif course_id is not None and question_id is not None:
+            external_image_capture_url = f"{server_canonical_host}/pl/course/{course_id}/question/{question_id}/variants/{html_params['variant_id']}/external_image_capture/answer/{answer_name}"
+        else:
+            raise ValueError(
+                "Either course_instance_id and instance_question_id or course_id and question_id must be available to use pl-image-capture."
+            )
+        html_params["external_image_capture_url"] = external_image_capture_url
```
It feels a bit weird and fragile that the element is expected to construct this URL for itself.

For workspaces, we pass in a full `workspace_url` in `options`:

https://github.com/PrairieLearn/PrairieLearn/blob/09b9386f2e32621310fb2f730435d555dec39870/apps/prairielearn/src/question-servers/freeform.ts#L1322

Ditto for things like `client_files_question_url` and `client_files_question_dynamic_url`. I'm thinking we should do the same here. Unlike the others, we can make it include the host as well instead of making folks manually add that to the front of the path.

nwalters512 commented on apps/prairielearn/src/lib/externalImageCapture.ts:
```diff
@@ -0,0 +1,46 @@
+    authn_user_id: resLocals.authn_user.authn_user_id,
+  });
+
+  // Create the ExternalImageCapture record
+  await sqldb.queryAsync(sql.insert_new_external_image_capture, {
+    variant_id: variantId,
+    answer_name: answerName,
+    file_id,
+  });
```
Talk to me about the design choices here. Why is a new database table needed? I'm wondering why we can't do something simpler like directly sending the image data over the websocket connection.

miguelaenlle replied:
I used a new database table to support the manual reload button, which depends on the data of the uploaded image being available and persisted for when the user clicks reload. 

I'd like to use a simpler design for transmitting images, such as directly sending image data over the websocket connection, as you suggested @nwalters512, though I have a few questions:

- Would you recommend continuing to include the manual reload button? If so, how would you recommend persisting the uploaded data? 
     - I added the manual reload button for the event that the uploaded image doesn't immediately appear, particularly if the user's assessment device has a network issue during the upload.
     
     - An approach to persisting the uploaded data I had in mind would be caching a URL to the uploaded image, so if the user clicks "Reload", and they uploaded an image recently, the backend will have a cache hit on the URL to the uploaded data and return it.
     - If we remove the manual reload button, what are your thoughts on having the external device wait for an "acknowledgement" message from the assessment device, and consider the request a failure if that acknowledgement is not received within an interval (e.g. 10 seconds)?

- Storing image data in the websocket: An uploaded image from my iPhone is 130KB large. Is it okay to transmit files of that size over websockets?

miguelaenlle replied:
Thank you for the suggestions, @nwalters512 . I'll remove the database table and the "Reload" button, and add a "Try again" button that appears in X seconds to the external image capture page. I'll also adjust this functionality such that the image data is transmitted directly through the websockets.

nwalters512 commented on apps/prairielearn/elements/pl-image-capture/info.json:
For lack of a better place to put this comment: I think we should cover ourselves and make it super explicit that neither this element nor its implementation strategy should be copied or forked. Maybe a `README.md` in this directory and a node in the Python file?

nwalters512 commented on apps/prairielearn/src/lib/externalImageCaptureSocket.ts:
```diff
@@ -0,0 +1,79 @@
+
+    socket.join(`variant-${msg.variant_id}-answer-${msg.answer_name}`);
+
+    const externalImageCaptureData: StatusMessage = {
+      variant_id: msg.variant_id,
+      answer_name: msg.answer_name,
+    };
+
+    callback(externalImageCaptureData);
```
Nit, avoiding a local variable where we can.

```suggestion
    callback({
      variant_id: msg.variant_id,
      answer_name: msg.answer_name,
    } satisfies StatusMessage);
```

nwalters512 commented on apps/prairielearn/src/lib/externalImageCaptureSocket.ts:
```diff
@@ -0,0 +1,79 @@
+    namespace
+      .to(`variant-${variant_id}-answer-${answer_name}`)
+      .emit('externalImageCapture', eventData);
+  } catch (err) {
+    logger.error('Error in emitExternalImageCapture', err);
+    Sentry.captureException(err);
+  }
```
Is this necessary? Can `namespace.emit(...)` ever fail?

miguelaenlle replied:
I confirmed that it's not, and removed the `emit`.

nwalters512 commented on apps/prairielearn/src/lib/externalImageCaptureSocket.types.ts:
AFAICT this is only used in a single file. Consider inlining it there?

nwalters512 commented on apps/prairielearn/src/migrations/20250602181038_external_image_capture__create.sql:
```diff
@@ -0,0 +1,14 @@
+  variant_id BIGINT NOT NULL REFERENCES variants ON DELETE CASCADE ON UPDATE CASCADE
+);
+
+ALTER TABLE external_image_capture
+DROP CONSTRAINT IF EXISTS external_image_capture_variant_id_and_answer_name_unique;
```
Is this necessary? When could the constraint already exist when this migration runs?

miguelaenlle replied:
As per this discussion https://github.com/PrairieLearn/PrairieLearn/pull/12061#discussion_r2129192378, I removed the `external_image_capture` table and migrations, so I'll resolve this thread.

nwalters512 commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.sql:
```diff
@@ -0,0 +1,19 @@
+  external_image_capture AS eic
+WHERE
+  eic.variant_id = $variant_id
+  AND eic.answer_name = $answer_name
+  -- BLOCK select_variant_by_id
```
Formatting:

```suggestion
  AND eic.answer_name = $answer_name;

-- BLOCK select_variant_by_id
```

nwalters512 commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,125 @@
+
+const sql = sqldb.loadSqlEquiv(import.meta.url);
+
+const router = express.Router({
+  mergeParams: true,
+});
```
Nit:

```suggestion
const router = express.Router({ mergeParams: true });
```

nwalters512 commented on apps/prairielearn/src/question-servers/freeform.ts:
Per my other comment, I'm hoping most/all of these can be replaced with something like `external_image_capture_url` that's pre-computed on the server.

nwalters512 reviewed (commented):
Here are some initial comments before I log off for the day. I didn't look to closely at the database schema or the route handlers since I wanted to hear about why those are necessary first (as opposed to something more stateless that just shuttles image data around over websockets).

miguelaenlle reviewed (commented):
Quick self-review.

eliotwrobson commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.py:
```diff
@@ -0,0 +1,118 @@
+    element = lxml.html.fragment_fromstring(element_html)
+    file_name = pl.get_string_attrib(element, "file-name")
+
+    submitted_file_content = data["submitted_answers"].get(file_name, None)
```
```suggestion
    submitted_file_content = data["submitted_answers"].get(file_name)
```

eliotwrobson commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.py:
```diff
@@ -0,0 +1,118 @@
+    file_name = pl.get_string_attrib(element, "file-name")
+    if not file_name.endswith(".png"):
+        pl.add_files_format_error(
+            data, f"File name '{file_name}' must end with '.png'."
```
```suggestion
            data, f"File '{file_name}' must have extension '.png'."
```
Slight wording nit.

eliotwrobson reviewed (commented):
Not super sure how some of the internals here work but a couple of minor comments.

jonatanschroeder commented on apps/prairielearn/assets/scripts/externalImageCaptureClient.ts:
```diff
@@ -0,0 +1,209 @@
+        // Perform scaling to ensure that user-uploaded images are not too large.
+        // The scale factor ensures that the width and height of the image do not exceed 1000px.
+        // If the image width and height are both less than 1000px, no scaling is applied.
+        const imageScaleFactor = 1000 / Math.max(image.width, image.height);
```
Should this 1000 be a constant at the top of the file?

jonatanschroeder replied:
Typical conventions for global constants is to use UPPERCASE_SNAKE_CASE. So you should use `MAX_IMAGE_SIDE_LENGTH`.

miguelaenlle replied:
Thank you for pointing this out, @jonatanschroeder . I made this change:

https://github.com/PrairieLearn/PrairieLearn/blob/06703e6c188c6ca1799b892ace9dade0633fd168/apps/prairielearn/assets/scripts/externalImageCaptureClient.ts#L10

jonatanschroeder commented on apps/prairielearn/assets/scripts/externalImageCaptureClient.ts:
```diff
@@ -0,0 +1,209 @@
+        const targetWidth = Math.round(image.width * imageScaleFactor);
+        const targetHeight = Math.round(image.height * imageScaleFactor);
+
+        const canvas = document.createElement('canvas');
+        canvas.width = targetWidth;
+        canvas.height = targetHeight;
+
+        const ctx = canvas.getContext('2d');
+        if (!ctx) {
+          throw new Error('Failed to get canvas context');
+        }
```
This is a bunch of processing that could be memoized instead of being repeated with every frame.

jonatanschroeder commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.mustache:
```diff
@@ -0,0 +1,123 @@
+    </p>
+
+    <div class="d-flex gap-1 mt-1">
+      <button class="js-capture-local-camera-image-button btn btn-info" disabled>
+        <i class="bi bi-camera-fill me-1"></i>
+        Capture image
+      </button>
+      <button class="js-cancel-local-camera-button btn btn-secondary">
+        Cancel
+      </button>
```
Including a [button type](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/button#type) is always a good practice. This is particularly important for buttons that are part of a form and that are not intended to be submit buttons, since that's the default for button types inside forms.

miguelaenlle replied:
Thank you for the feedback, @jonatanschroeder . I added button types to all of the buttons in the PR 👍

jonatanschroeder commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.js:
```diff
@@ -0,0 +1,556 @@
+      });
+
+      captureLocalCameraImageButton.addEventListener('click', (event) => {
+        event.preventDefault();
```
I believe adding a `type="button"` to the corresponding button will allow you to get rid of the `preventDefault()`.

miguelaenlle replied:
Got it -- I added a `type` to all the buttons in the `pl-image-capture` element and removed the `event.preventDefault()` calls 👍

jonatanschroeder commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.py:
```diff
@@ -0,0 +1,118 @@
+        element, "mobile-capture-enabled", MOBILE_CAPTURE_ENABLED_DEFAULT
+    )
+
+    if data["panel"] != "question":
+        return ""
```
Should this show a preview of the capture in the submission panel?

jonatanschroeder commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.py:
```diff
@@ -0,0 +1,118 @@
+        return
+
+    # Confirm that the data URI is a PNG image.
+    if not submitted_file_content.startswith("data:image/png;base64,"):
+        pl.add_files_format_error(
+            data, f"Image submission for {file_name} is not a PNG image."
+        )
+        return
```
Nit: this is not foolproof to detect PNG format, this is only taking into account the reported mimetype validated client-side. We already have a Pillow dependency, so we could use it to detect the image type from the content itself. Won't block on it, though.

jonatanschroeder replied:
From what I could gather, the [capture mechanism](https://w3c.github.io/html-media-capture/#dfn-capture) does not need to honour the format of `accept` other than distinguishing between `image/*`, `video/*`, `audio/*`, etc. So you may not receive a PNG by default (and to be honest, I would be surprised if any capture device decides to use PNG by default, since JPEG tends to be more efficient for pictures than PNG).

One option you have is to accept a larger range of images (e.g., JPEG, PNG, [WebP](https://developer.mozilla.org/en-US/docs/Web/Media/Guides/Formats/Image_types#webp_image)), and try to [convert the image](https://pillow.readthedocs.io/en/latest/reference/Image.html#PIL.Image.Image.convert) to an appropriate type before saving if the image is supported by Pillow but not accepted in our list.

jonatanschroeder replied:
This might also be an artifact of the canvas resizing process. Check if adding an explicit [`type` parameter to `toBlob()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toBlob#type) may affect the result. If so, you could use the canvas both to resize and to convert the image.

BTW, I'd encourage you to shift away from PNG towards JPG or WEBP if you are worried about file size.

jonatanschroeder commented on apps/prairielearn/src/ee/lib/ai-grading.ts:
```diff
@@ -196,6 +196,7 @@ export async function aiGrade({
           response_format: zodResponseFormat(RubricGradingResultSchema, 'score'),
           temperature: OPEN_AI_TEMPERATURE,
         });
+
```
I think this snuck in?

miguelaenlle replied:
Yup, it did. I fixed this 👍

jonatanschroeder commented on apps/prairielearn/src/lib/question-render.ts:
```diff
@@ -15,12 +15,12 @@ import {
 } from '../components/QuestionContainer.types.js';
 import { QuestionNavSideButton } from '../components/QuestionNavigation.html.js';
 import { QuestionScorePanelContent } from '../components/QuestionScore.html.js';
+import type { SubmissionForRender } from '../components/SubmissionPanel.html.js';
 import {
   SubmissionBasicSchema,
   SubmissionDetailedSchema,
   SubmissionPanel,
 } from '../components/SubmissionPanel.html.js';
```
Nit: combine these imports?
```suggestion
import {
  SubmissionBasicSchema,
  SubmissionDetailedSchema,
  type SubmissionForRender,
  SubmissionPanel,
} from '../components/SubmissionPanel.html.js';
```

jonatanschroeder commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,76 @@
+    const variantId = req.params.variant_id as string;
+    const fileName = req.query.file_name as string;
+
+    const variant = await selectAndAuthzVariant({
+      unsafe_variant_id: variantId,
+      variant_course: res.locals.course,
+      question_id: res.locals.question.id,
+      course_instance_id: res.locals?.course_instance?.id,
+      instance_question_id: res.locals.instance_question?.id,
+      authz_data: res.locals.authz_data,
+      authn_user: res.locals.authn_user,
+      user: res.locals.user,
+      is_administrator: res.locals.is_administrator,
+      publicQuestionPreview: res.locals.public_question_preview,
+    });
+
+    if (!variant) {
+      throw new HttpStatusError(404, 'Variant not found');
+    }
+
+    if (!variant.open) {
+      throw new HttpStatusError(403, 'This variant is not open');
+    }
```
This validation should probably be done in the GET as well, or in a `router.use` before it.

jonatanschroeder commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,76 @@
+  '/',
+  asyncHandler(async (req, res) => {
+    const variantId = req.params.variant_id as string;
+    const fileName = req.query.file_name as string;
```
Nit: I'm a bit wary of using `req.query` in POST. Can the file name be moved to a hidden input and use a `req.body` instead?

miguelaenlle replied:
Yes -- I moved `file_name` into a hidden input and used `req.body.file_name` instead:

https://github.com/PrairieLearn/PrairieLearn/blob/bd350646e22f572528048c087a3432200e250121/apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts#L58-L67

jonatanschroeder commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    // Emit a socket event to notify the client that the image has been captured.
+    await emitExternalImageCapture({
+      variant_id: variantId,
+      file_name: fileName,
+      file_content: req.file.buffer.toString('base64'),
+    });
```
For consideration: this is potentially a pretty big file to send using web sockets. Since the data is in the server anyway, would it be better to save the file in a short-lived cache (e.g., for 30 minutes at most), then send the notification of the file existing in the socket, including a random (UUID-based? hash-based?) URL to retrieve it?

jonatanschroeder commented on apps/prairielearn/src/server.ts:
```diff
@@ -258,10 +266,18 @@ export async function initExpress(): Promise<Express> {
     '/pl/course_instance/:course_instance_id(\\d+)/instructor/question/:question_id(\\d+)/file_view/*',
     upload.single('file'),
   );
+  app.use(
```
I believe this should be `app.post`, right?

miguelaenlle replied:
Yes, I made this change:

https://github.com/PrairieLearn/PrairieLearn/blob/bd350646e22f572528048c087a3432200e250121/apps/prairielearn/src/server.ts#L273-L276

jonatanschroeder reviewed (commented):
This is just a first skim over the code, I haven't tested it.

reteps commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.js:
```diff
@@ -0,0 +1,554 @@
+      }
+
+      captureWithLocalCameraButton.addEventListener('click', () => {
+        this.startLocalCameraCapture();
```
Ah, I see, thanks.

miguelaenlle commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.js:
```diff
@@ -0,0 +1,554 @@
+     */
+    listenForExternalImageCapture() {
+      const questionContainer = document.querySelector('.question-container');
+      if (!questionContainer) return;
```
`questionContainer` contains `variantToken`, which was originally required for joining the external image capture socket:

https://github.com/PrairieLearn/PrairieLearn/blob/47fd3939eeffae850b31612a8378f5f1c6f29484/apps/prairielearn/elements/pl-image-capture/pl-image-capture.js#L106-L123

However, I removed this requirement, since external image capture page, which does not have a `variantToken` available, now also listens to the WebSocket to receive an acknowledgement message that the image has been loaded successfully. 

I have since removed this check for `questionContainer` since it's no longer needed.

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,87 @@
+      throw new HttpStatusError(400, 'No file uploaded');
+    }
+
+    if (req.file.buffer.length > 10 * 1024 * 1024) {
```
Yes, that is correct. I added this check to handle the edge case that a user somehow uploads a very massive image to PL's web server, since resizing occurs on the clientside.

miguelaenlle commented on apps/prairielearn/assets/scripts/externalImageCaptureClient.ts:
```diff
@@ -0,0 +1,220 @@
+
+  cameraInput.addEventListener('change', () => {
+    cameraInput.disabled = false;
+    if (cameraInput.files && cameraInput.files.length > 0) {
```
Agreed that this makes the code more readable. I made this change:

https://github.com/PrairieLearn/PrairieLearn/blob/07551d990d84c8b451f1de38ce0001a31ba59594/apps/prairielearn/assets/scripts/externalImageCaptureClient.ts#L150-L163

miguelaenlle commented on apps/prairielearn/assets/scripts/externalImageCaptureClient.ts:
```diff
@@ -0,0 +1,220 @@
+          }
+
+          const resizedFile = new File([blob], file.name, { type: file.type });
+          const dt = new DataTransfer();
```
I used `DataTransfer` to transfer the resized file data into `cameraInput.files`. However, I was able to reimplement this using the data URL of `resizingCanvas` directly:

https://github.com/PrairieLearn/PrairieLearn/blob/07551d990d84c8b451f1de38ce0001a31ba59594/apps/prairielearn/assets/scripts/externalImageCaptureClient.ts#L190-L199

miguelaenlle replied:
Reopening this since, after further testing, I discovered that the images aren't being properly resized to the target width/height due to my implementation. I'll post in this thread once this issue is resolved.

miguelaenlle replied:
I reverted the change above and switched back to using `DataTransfer`, since the image wasn't properly resized without the `toBlob` call:

https://github.com/PrairieLearn/PrairieLearn/blob/65b506d9a766ff91f5ac5604651019b94868aa74/apps/prairielearn/assets/scripts/externalImageCaptureClient.ts#L184-L207

Also, I couldn't find a better/more intuitive way to update `files`, and converting `blob` to a data URL is also quite complicated.

miguelaenlle commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.js:
```diff
@@ -0,0 +1,554 @@
+      if (qrCode) {
+        qrCode.innerHTML = qrCodeSvg;
+      } else {
+        throw new Error('QR code element not found.');
```
I made this change:

https://github.com/PrairieLearn/PrairieLearn/blob/07551d990d84c8b451f1de38ce0001a31ba59594/apps/prairielearn/elements/pl-image-capture/pl-image-capture.js#L200-L213

miguelaenlle replied:
I made this change:

https://github.com/PrairieLearn/PrairieLearn/blob/af1d8e36fd81c42c94bd9682efb02ed93c5df2ad/apps/prairielearn/elements/pl-image-capture/pl-image-capture.js#L200-L211

miguelaenlle commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.js:
```diff
@@ -0,0 +1,554 @@
+      });
+    }
+
+    setNoCaptureAvailableYetState(uploadedImageContainer) {
```
I added a JSDoc comment and removed the check for `uploadedImageContainer`. Similarly, `setLoadingCaptureState` was called after a validation that it exists as well, so I removed the check for `uploadedImageContainer` there as well:

https://github.com/PrairieLearn/PrairieLearn/blob/af1d8e36fd81c42c94bd9682efb02ed93c5df2ad/apps/prairielearn/elements/pl-image-capture/pl-image-capture.js#L266-L297

miguelaenlle commented on apps/prairielearn/src/tests/accessibility/index.test.ts:
```diff
@@ -342,6 +342,11 @@ const SKIP_ROUTES = [
   '/pl/course_instance/:course_instance_id/instructor/instance_question/:instance_question_id/file/:filename',
   '/pl/course_instance/:course_instance_id/instructor/instance_question/:instance_question_id/text/:filename',
+
+  '/pl/course/:course_id/question/:question_id/externalImageCapture/variant/:variant_id',
```
Yup! I added this update;

https://github.com/PrairieLearn/PrairieLearn/blob/626a3c0648363f9c952a32f6a59e0da6fe017556/apps/prairielearn/src/tests/accessibility/index.test.ts#L292-L295

miguelaenlle commented on testCourse/questions/demo/imageCapture/info.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "uuid": "ef9df978-d93e-4710-aef6-dcf7eae4cd3c",
+  "title": "Image capture demo",
```
Sounds good; thank you for the clarification, @jonatanschroeder . I'll keep this question in `testCourse`, since I'm using this to test interaction between more than one capture element, and I'll also create a new question demonstrating the various settings of `pl-image-capture` in `exampleCourse`.

miguelaenlle replied:
I added an example and template question to `exampleCourse` for the `pl-image-capture` element. More details are available in this comment: https://github.com/PrairieLearn/PrairieLearn/pull/12061#issuecomment-2974002335

reteps commented on apps/prairielearn/elements/pl-image-capture/README.md:
```diff
@@ -0,0 +1,5 @@
+# PrairieLearn Image Capture Element
```
The documentation on `elements.md` should be updated as well.

In that new documentation, make sure to add a warning that the QR code feature requires both devices to be on the same network when doing local development.

jonatanschroeder replied:
> In that new documentation, make sure to add a warning that the QR code feature requires both devices to be on the same network when doing local development.

Since that documentation is for instructors, not developers, I think it's better not to encourage instructors to attempt the QR code functionality, since it requires them to also set up `config.serverCanonicalHost`. One option is to have a message, on a local installation, whenever `serverCanonicalHost` is null, that "mobile device functionality is not available in this environment", or something similar.

reteps replied:
I like that suggestion @jonatanschroeder . Perhaps a note in dev setup for serverCanonicalHost in addition?

miguelaenlle replied:
I added the note in dev setup. More details are in this comment: https://github.com/PrairieLearn/PrairieLearn/pull/12061#issuecomment-2974002335

miguelaenlle replied:
Also, I added an explanation of the `pl-image-capture` element to `elements.md`:

![Screenshot 2025-06-15 at 9 55 28 AM](https://github.com/user-attachments/assets/780baffd-7964-4ba4-93d9-d2498e58f496)

![Screenshot 2025-06-15 at 9 55 46 AM](https://github.com/user-attachments/assets/1a528222-dc99-412e-ba6f-46a307c32927)

nwalters512 commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.js:
```diff
@@ -0,0 +1,554 @@
+        throw new Error(`Image capture element with UUID ${uuid} not found.`);
+      }
+
+      const options = JSON.parse(this.imageCaptureDiv.dataset.options);
```
The thought has been at the back of my mind for ages, yes. Historically we have kept the relatively-useful property that someone can directly copy something out of `elements/` into their own course to modify it. It would be very nice to somehow keep that property if/when we do this, though we can also talk through if that's desired.

nwalters512 commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,87 @@
+      authn_user: res.locals.authn_user,
+      user: res.locals.user,
+      is_administrator: res.locals.is_administrator,
+      publicQuestionPreview: res.locals.public_question_preview,
```
👀 two questions:

- Will remote capture work on public preview pages?
- _Should_ it work on public preview pages?

miguelaenlle replied:
I tested remote capture on public preview pages, and it currently doesn't work, but I can make it work by adding a `pl/public/course/:course_id/question/:question_id/externalImageCapture/...` endpoint, similar to what was added for `pl/course/:course_id/question/:question_id/...`. 

Regarding if we should support it: I think it would be beneficial to support it, since it's not too hard to add and would enable us to add a question with `pl-image-capture` to the PrairieLearn Question Catalog. 

Regarding security, if I secure the WebSockets with the question's `variantToken`, I think that would resolve the vulnerability described in this discussion: https://github.com/PrairieLearn/PrairieLearn/pull/12061#discussion_r2138859876

miguelaenlle replied:
I implemented functionality for remote capture on public preview pages:

![Screenshot 2025-06-15 at 10 42 39 AM](https://github.com/user-attachments/assets/29d3a248-7cf7-4f78-8c51-cf2feb25284d)

![Screenshot 2025-06-15 at 10 43 05 AM](https://github.com/user-attachments/assets/c2238241-91e4-4735-8dbf-48428b39105a)

nwalters512 reviewed (commented):
Another round of comments! Thanks for the patience with all the rounds of reviews. This is complex stuff, but we're on the right track here 🙌

miguelaenlle commented on apps/prairielearn/assets/scripts/externalImageCaptureClient.ts:
I think that Preact would clean up the implementation of the external image capture page, but it would only simplify the pagination logic. The image resizing and WebSocket logic would be very similar. 

Preact would probably be more beneficial if the upcoming crop and rotation functionality is implemented in the external image capture page. This would improve the user experience, enabling users to crop/rotate immediately after they capture the image, but would also introduce complexity through (1) requiring rotation/crop functionality to have robust mobile support and (2) potentially resulting in two repeated implementations of crop/rotation functionality here and in `pl-image-capture` for cropping/rotating images captured on a local camera. 

@nwalters512 , what are your thoughts on implementing crop/rotation in both the external image capture page and `pl-image-capture`, or only implementing it in `pl-image-capture`? (CC @mwest1066)

mwest1066 replied:
I think that having crop/rotate in `pl-image-capture` is the essential thing. My initial thought is that we don't also need it in the external page, but this is something where user testing would be good to see whether people try to crop already in the external page when taking the photo.

miguelaenlle replied:
Thank you for the feedback, @mwest1066 . In a subsequent PR I'll implement crop/rotation only in `pl-image-capture`, and we can add it to external image capture if that's desirable for users.

miguelaenlle commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.js:
Would it be possible/recommendable to use Preact in the `pl-image-capture` component?

miguelaenlle commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.js:
```diff
@@ -0,0 +1,567 @@
+
+      this.external_image_capture_url = options.external_image_capture_url;
+      this.submitted_file_name = options.submitted_file_name;
+      this.submission_date = options.submission_date;
```
`submission_date` isn't used anywhere; this was left over from the reload function. Apologies about that, and I removed the variable.

miguelaenlle commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.js:
```diff
@@ -0,0 +1,567 @@
+        await localCameraVideo.play();
+
+        // Hide the permission message
+        permissionMessage.classList.add('d-none');
```
I adjusted the functionality to display a message only when an error occurs, and to change the message based on the error type:

- The message flash no longer occurs:

https://github.com/user-attachments/assets/c5fd4eae-fb72-48fe-bd2e-e96d5e995600

- Different error messages:

![Screenshot 2025-06-11 at 2 52 57 PM](https://github.com/user-attachments/assets/fbdf3c5f-d7af-4f6e-8dc4-f06e5e63a193)

![Screenshot 2025-06-11 at 2 49 49 PM](https://github.com/user-attachments/assets/c6f87def-2438-43e9-9c26-93c101c55f02)

![Screenshot 2025-06-11 at 2 45 45 PM](https://github.com/user-attachments/assets/c4f7bfa2-ba94-46aa-8754-735a3898a676)

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.html.ts:
```diff
@@ -0,0 +1,101 @@
+              </label>
+
+              <p class="text-muted mt-3">
+                Before uploading, ensure that your entire solution is visible, legible, and well-lit
```
I adjusted the text here and in the `pl-image-capture` element:

https://github.com/PrairieLearn/PrairieLearn/blob/76521f1a926cb8e08f79d7bc74170924049f9fb6/apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.html.ts#L88-L91

https://github.com/PrairieLearn/PrairieLearn/blob/76521f1a926cb8e08f79d7bc74170924049f9fb6/apps/prairielearn/elements/pl-image-capture/pl-image-capture.mustache#L87-L89

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,87 @@
+
+import { ExternalImageCapture } from './externalImageCapture.html.js';
+
+const router = express.Router({ mergeParams: true });
```
Yes, I made this change:

d7bc74170924049f9fb6/apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts#L1-L11

miguelaenlle commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.py:
```diff
@@ -0,0 +1,138 @@
+    pl.check_attribs(
+        element,
+        required_attribs=["file-name"],
+        optional_attribs=["mobile-capture-enabled"],
```
As discussed over Zoom, I'll keep mobile capture enabled by default, since it would be needed for the majority of questions, and the QR code would be unusable anyways in an environment where students don't have a mobile device (e.g. the CBTF).

miguelaenlle commented on apps/prairielearn/src/lib/externalImageCaptureSocket.ts:
```diff
@@ -0,0 +1,76 @@
+}
+
+export function connection(socket: Socket) {
+  socket.on('joinExternalImageCapture', async (msg, callback) => {
```
As per our discussion on Zoom, I'll use `variantToken` as described above to secure the websocket.

miguelaenlle replied:
I implemented this functionality. The `variant_token` is now required to connect to the external image capture socket. 

https://github.com/PrairieLearn/PrairieLearn/blob/747deb6f32c68b7738db6edc5a593ef5592b1fcd/apps/prairielearn/src/lib/externalImageCaptureSocket.ts#L20-L54

https://github.com/PrairieLearn/PrairieLearn/blob/747deb6f32c68b7738db6edc5a593ef5592b1fcd/apps/prairielearn/src/lib/externalImageCaptureSocket.ts#L91-L122

I also added `file_name` and `variant_id` validation as per this CodeRabbit recommendation: https://github.com/PrairieLearn/PrairieLearn/pull/12061#discussion_r2147380955 . Let me know your thoughts on including this.

miguelaenlle commented:
@reteps , I implemented the mobile capture suggestions listed above:

> For the mobile image capture:
> 
> * move the upload and retake to the same line, and remove the dashed blue box.

Here's what the external image capture flow looks like now:

![IMG_2618](https://github.com/user-attachments/assets/b26fbc01-57d9-4d77-a25e-79a244770ac7)

![IMG_2619](https://github.com/user-attachments/assets/fb03a68f-318f-4bed-a22c-cdb7a2379254)

> * on the success page, include a button to retake. My browser didn't auto-refresh off this page upon a recapture, and I was 'stuck' there.

Now, on the success page, the user can click "Retake Photo". This opens their camera immediately, and repeats the flow shown above (take photo -> confirm or retake -> upload -> success page)

![IMG_2620](https://github.com/user-attachments/assets/2e1caaca-e31a-4cf6-a175-49b95a8aea0e)

If the upload fails, the user is prompted with a "Try again" button:

![IMG_2621](https://github.com/user-attachments/assets/eb7f88d0-3796-4f3c-9887-89da279a3e82)

This brings them back to the image confirmation/retake page, and the "Upload" button now says "Reupload": 

![IMG_2622](https://github.com/user-attachments/assets/431f0e01-db9f-4ac5-bd1b-ac5b169d04d6)

Alternatively, the user can retake the image at this point.

miguelaenlle commented:
I also implemented the other suggestions, @reteps :

> Other:
> 
> * Example within example course

I added a template question and an `element` sample question:

![Screenshot 2025-06-15 at 9 35 47 AM](https://github.com/user-attachments/assets/63fd0623-57a9-48c9-b67e-e564ac14df9f)

![Screenshot 2025-06-15 at 9 37 07 AM](https://github.com/user-attachments/assets/4620b60b-a74d-4a93-88e2-a395f27e995c)

To ensure `exampleCourseQuestions.test.ts` passes, I also implemented testing for `pl-image-capture`:

https://github.com/PrairieLearn/PrairieLearn/blob/56895f62ecf6086f75d75cdef1db4cfd27f59c31/apps/prairielearn/elements/pl-image-capture/pl-image-capture.py#L137-L169

> * Note in dev setup about how to get this element to work

I added a note in `configJson.md`:

https://github.com/PrairieLearn/PrairieLearn/blob/947c17c5a2ea4359eb641fb340f98a3a3e349c4f/docs/dev-guide/configJson.md?plain=1#L1-L61

> * Warning if `serverCanonicalHost` isn't set / mention

If `serverCanonicalHost` is `null` in `config.json~, if the user attempts to scan a QR code, this appears:

![Screenshot 2025-06-15 at 9 50 43 AM](https://github.com/user-attachments/assets/23f40b50-e71a-48c1-9148-dec632cabf25)

miguelaenlle commented on apps/prairielearn/src/lib/externalGradingSocket.ts:
```diff
@@ -39,7 +38,7 @@ export function connection(socket: Socket) {
       return callback(null);
     }
-    if (!checkToken(msg.variant_token, msg.variant_id)) {
+    if (!checkVariantToken(msg.variant_token, msg.variant_id)) {
```
I made this change, since `checkVariantToken` is also used in `externalImageCaptureSocket.ts`.

reteps commented on docs/elements.md:
```diff
@@ -383,6 +384,47 @@ The `pl-file-upload` element and the contents of the uploaded file(s) are only d
+
+In manual grading mode, staff can view submitted images in the submission panel and, if added, through the [`pl-file-preview`](#pl-file-preview-element) element, where submitted images will appear under their associated `file-name`.
+
+By default, the `mobile-capture-enabled` setting is `true`. We strongly recommend keeping mobile capture enabled for most questions to allow students to capture high-quality images easily.
```
Link to the dev guide configJson here as well.

reteps reviewed (approved):
Everything looks good to me -- nice work @miguelaenlle !

nwalters512 commented on apps/prairielearn/elements/pl-image-capture/pl-image-capture.py:
```diff
@@ -0,0 +1,170 @@
+    elif result == "invalid":
+        data["raw_submitted_answers"][file_name] = ""
+
+        if not data["format_errors"]:
+            data["format_errors"] = {}
```
Nit, I believe this is unnecessary, this dictionary should always be present.

miguelaenlle replied:
Yup, I removed this dictionary:

https://github.com/PrairieLearn/PrairieLearn/blob/6ccac26ff2ddd4b7f91bf7f75845344c67675b01/apps/prairielearn/elements/pl-image-capture/pl-image-capture.py#L159-L167

nwalters512 reviewed (commented):
Overall this is looking very good! I'm relying on review from Peter and others for much of the element code. I focused my attention on the server bits and the docs/examples.

miguelaenlle commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,98 @@
+      authn_user: res.locals.authn_user,
+      user: res.locals.user,
+      is_administrator: res.locals.is_administrator,
+      publicQuestionPreview: res.locals.public_question_preview,
```
I must have accidentally deleted this at some point. I added it here: 

https://github.com/PrairieLearn/PrairieLearn/blob/6ccac26ff2ddd4b7f91bf7f75845344c67675b01/apps/prairielearn/src/server.ts#L1781-L1787

miguelaenlle commented on exampleCourse/courseInstances/SectionA/assessments/questionTemplates/infoAssessment.json:
```diff
@@ -19,7 +19,8 @@
         { "id": "template/symbolic-input/random", "points": 2, "maxPoints": 10 },
-        { "id": "template/figure/dynamic", "points": 1, "maxPoints": 4 }
+        { "id": "template/figure/dynamic", "points": 1, "maxPoints": 4 },
+        { "id": "template/image-capture/static", "points": 1, "maxPoints": 1 }
```
I removed it in 6ccac26ff2ddd4b7f91bf7f75845344c67675b01

nwalters512 commented on testCourse/questions/demo/imageCapture/question.html:
```diff
@@ -0,0 +1,17 @@
+<pl-question-panel>
+    <p>Consider the function $f(x) = x^3+2x^2+x+1$. Draw the graphs of the following functions: </p>
```
Nit, consistent indentation:

```suggestion
  <p>Consider the function $f(x) = x^3+2x^2+x+1$. Draw the graphs of the following functions: </p>
```

nwalters512 commented on exampleCourse/questions/template/image-capture/static/question.html:
```diff
@@ -0,0 +1,3 @@
+<p>Consider the function $f(x) = arcsin(2x) + cos(4x^3-5x)+e^{x^3}$. Find $f'(x)$ and show your work.</p>
+<pl-image-capture file-name="solution.jpg"></pl-image-capture> 
+
```
Nit, extraneous newline.

```suggestion
```

nwalters512 commented on exampleCourse/questions/element/imageCapture/question.html:
```diff
@@ -0,0 +1,23 @@
+<pl-submission-panel>
+  <pl-file-preview></pl-file-preview>
+</pl-submission-panel>
+
+<pl-answer-panel>
+</pl-answer-panel>
+
```
This isn't needed:

```suggestion
```

Note that when I talk about a trailing newline, that just means a single newline after the last actual line of content. There are actually two newlines here now!

nwalters512 commented on exampleCourse/questions/element/imageCapture/question.html:
```diff
@@ -0,0 +1,23 @@
+<pl-question-panel>
+<p>This question demonstrates the image capture element, which enables students to submit images to questions using their local camera or an external device (e.g. a phone or tablet).</p>
```
Nit for consistent indentation (sorry we don't have tooling to handle this yet!)

```suggestion
  <p>This question demonstrates the image capture element, which enables students to submit images to questions using their local camera or an external device (e.g. a phone or tablet).</p>
```

nwalters512 commented on exampleCourse/questions/element/imageCapture/question.html:
```diff
@@ -0,0 +1,23 @@
+</pl-question-panel> 
+
+<pl-card>
+    <p>Consider the function $f(x) = arcsin(2x) + cos(4x^3-5x)+e^{x^3}$. Find $f'(x)$ and show your work.</p>
+    <pl-image-capture file-name="solution-1.jpg"></pl-image-capture>
```
```suggestion
  <p>Consider the function $f(x) = arcsin(2x) + cos(4x^3-5x)+e^{x^3}$. Find $f'(x)$ and show your work.</p>
  <pl-image-capture file-name="solution-1.jpg"></pl-image-capture>
```

nwalters512 commented on exampleCourse/questions/element/imageCapture/question.html:
```diff
@@ -0,0 +1,23 @@
+<p>If necessary, mobile capture can be disabled.</p>
+
+<pl-card>
+    <p>Sketch the bounded region enclosed by the curves $y=-x^2+3$ and $y=2x^2$.</p>
+    <pl-image-capture file-name="solution-2.jpg" mobile-capture-enabled="false"></pl-image-capture>
```
```suggestion
  <p>Sketch the bounded region enclosed by the curves $y=-x^2+3$ and $y=2x^2$.</p>
  <pl-image-capture file-name="solution-2.jpg" mobile-capture-enabled="false"></pl-image-capture>
```

nwalters512 commented on apps/prairielearn/src/server.ts:
```diff
@@ -1740,6 +1778,14 @@ export async function initExpress(): Promise<Express> {
   );
+  app.use(
+    '/pl/public/course/:course_id(\\d+)/question/:question_id(\\d+)/externalImageCapture/variant/:variant_id(\\d+)',
+    (req, res, next) => {
+      res.locals.public_question_preview = true;
+      next();
+    },
+    (await import('./pages/externalImageCapture/externalImageCapture.js')).default,
```
Nit, @SethPoulsen has taken a different approach with handlers that need to serve both public and private routes:

https://github.com/PrairieLearn/PrairieLearn/blob/fb9d6ddb8e05be4cc926eb635c14a69258f223b8/apps/prairielearn/src/server.ts#L1747-L1753

Could we do something similar here for consistency?

miguelaenlle replied:
Yes. I implemented a similar handler here and removed the middleware:

https://github.com/PrairieLearn/PrairieLearn/blob/237fa72411727ee7f7d7c7f8bdaeb9fc00b3d8ba/apps/prairielearn/src/server.ts#L1777-L1782

https://github.com/PrairieLearn/PrairieLearn/blob/237fa72411727ee7f7d7c7f8bdaeb9fc00b3d8ba/apps/prairielearn/src/server.ts#L1427-L1430

nwalters512 commented on apps/prairielearn/src/server.ts:
```diff
@@ -1401,6 +1424,13 @@ export async function initExpress(): Promise<Express> {
     '/pl/course_instance/:course_instance_id(\\d+)/instance_question/:instance_question_id(\\d+)',
     (await import('./pages/studentInstanceQuestion/studentInstanceQuestion.js')).default,
   );
+  // External image capture page for an instance question. Enables users to capture and submit
+  // images from an external device (e.g. a mobile phone).
```
Nit: IMO the comment is unnecessary, we don't typically add comments like this in `server.ts`.

```suggestion
```

nwalters512 commented on apps/prairielearn/src/server.ts:
```diff
@@ -1504,6 +1534,14 @@ export async function initExpress(): Promise<Express> {
     next();
   });
+
+  // External image capture page for a question variant generated on the question preview page. Enables
+  // users to capture and submit images from an external device (e.g. a mobile phone).
```
Nit:

```suggestion
```

nwalters512 commented on apps/prairielearn/src/question-servers/freeform.ts:
```diff
@@ -914,6 +914,10 @@ async function renderPanel(
     ),
     submission_files_url: submission ? submissionFilesUrl : null,
+
+    variant_id: variant?.id,
```
Nit, `variant` should always be defined so optional chaining isn't necessary and could in fact be confusing:

```suggestion
    variant_id: variant.id,
```

nwalters512 commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,98 @@
+    // Emit a socket event to notify the client that the image has been captured.
+    emitExternalImageCapture({
+      variant_id: variantId,
+      variant_token: res.locals.variantToken,
```
Is it necessary to provide the token here?

miguelaenlle replied:
I think it makes sense to remove the token here, since it's not processed in the `emitExternalImageCapture` function and is never used when it's sent to the client. 

Moreover, the only way one could receive a message from `emitExternalImageCapture` is connecting via `joinExternalImageCapture`, and that requires having a valid `variantToken`.

nwalters512 commented on apps/prairielearn/src/pages/externalImageCapture/externalImageCapture.ts:
```diff
@@ -0,0 +1,98 @@
+);
+
+// Handles image uploading from the external image capture page.
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
```
Do you need a `selectAndAuthzVariant(...)` call here? Otherwise it seems like a malicious user could post to this route with an arbitrary `variant_id` and `file_name`.

nwalters512 reviewed (commented):
My primary remaining concern is the possible lack of authorization in the `POST` handler. Pretty much everything else is just nits.

nwalters512 reviewed (approved):
Let's ship this!
